### ISSUE TITLE
Strange side effect with SetKeyboardFocusHere()

### ISSUE THREAD

**Issue by JPGygax68** [ocornut/imgui#4054]
I just noticed that when I'm using `SetKeyboardFocusHere()` to set the focus on an `InputText()`, I can no longer drag other windows with the mouse (tried with the Demo window).

My situation is a bit peculiar because I'm using an ImGui window that uses the full client area of the SDL window, on which I've set the flag `ImGuiWindowFlags_NoBringToFrontOnFocus`.

I can get rid of the effect by applying a condition:
```C++
if (!ImGui::IsAnyItemActive()) 
    ImGui::SetKeyboardFocusHere();
```

I'll try to simplify the code so that it can be pasted into a main loop:
```c++
ImGuiIO& io = ImGui::GetIO();
static ImGuiWindowFlags flags = ImGuiWindowFlags_NoDecoration | ImGuiWindowFlags_NoMove | ImGuiWindowFlags_NoResize | ImGuiWindowFlags_NoSavedSettings | ImGuiWindowFlags_NoBringToFrontOnFocus;
ImGui::SetNextWindowPos(ImVec2(0, 0));
ImGui::SetNextWindowSize(io.DisplaySize);

ImGui::PushStyleVar(ImGuiStyleVar_WindowPadding, { ImGui::GetFontSize() * 4, ImGui::GetFontSize() * 4 });
auto open = ImGui::Begin("Full-screen window", 0, flags);
ImGui::PopStyleVar();
if (open) {
    ImGui::AlignTextToFramePadding(); ImGui::Text("Select function:"); ImGui::SameLine();
    ImGui::SetNextItemWidth(ImGui::GetFontSize() * 2);
    if (!ImGui::IsAnyItemActive()) // <--- without this condition, other windows will be unmovable and unreactive
        ImGui::SetKeyboardFocusHere();
    static char func_buf[3] = {};
    (void) ImGui::InputText("###Function", func_buf, sizeof(func_buf), ImGuiInputTextFlags_CharsDecimal);
}
ImGui::End();
```

**Comment by rokups** [comment]
`SetKeyboardFocusHere()` sets focused window. If you call it unconditionally each time you focus another window - focus is immediately reset. Other windows can not be dragged because their focus is broken.

What are you trying to achieve exactly?

If you need to set initial input focus you can use `ImGui::SetItemDefaultFocus()`.

If you need to have `InputText()` focused while allowing interaction with other widgets you could keep using your version with if condition. Maybe add `ImGui::IsWindowFocused()` for a good measure too.

**Comment by JPGygax68** [comment]
Thanks! In this case, it's more about understanding than about achieving something. The extra windows would not be there in a finished application anyway.

`ImGui::SetItemDefaultFocus()` should do what I was aiming for, but appears to do nothing. (I stepped into it with the debugger, the `g.NavInitXXX` variables are being set, but no caret appears in the input field.)

**Comment by rokups** [comment]
Make sure you place `ImGui::SetItemDefaultFocus()` _after_ item you want to be default-focused. Default focus is only applied when window is appearing (was not rendered on previous frame, but is rendered on current frame) and wont work when window regains focus after being merely in a background.

**Comment by ocornut** [comment]
It's unclear what you are trying to "solve" by removing the `if (!ImGui::IsAnyItemActive())` test which is correct ihmo.

SetKeyboardFocusHere() and SetItemDefaultFocus() live in different worlds...
I'll try to explain.

`SetKeyboardFocusHere()` is an old, misnamed function, needing a refactor and a better version of it may be called, e.g. `ActivateItem()`. In practice is both sets focus and activate the input text. The transition isn't so simple in reality because `SetKeyboardFocusHere()` supports a offset-kludge to allow activating sub-fields of a multi-field widget., which makes it not as easy to rework in a new api. 

`SetItemDefaultFocus()` sets item focus but currently that only visually reflected if navigation is being used (e.g. done via keyboard/gamepad), as we hide the navigation rectangle when using mouse.

**Comment by JPGygax68** [comment]
Thank you Omar & Rokas,

`SetItemDefaultFocus()` doesn't appear to work even when putting it after the `InputText()`.

@ocornut Why not provisionally implement that function `ActivateItem()` by combining `SetKeyboardFocusHere()` with `IsAnyItemActive()` and possibly `IsWindowFocused()`?

For now, of course, I'll stick to your solution.

**Comment by ocornut** [comment]
> Why not provisionally implement that function ActivateItem() by combining SetKeyboardFocusHere() with IsAnyItemActive() and possibly IsWindowFocused()?

`ActivateItem()` would NOT call IsAnyItemActive() or IsWindowFocused(), that sentence suggest you might not understand Rokas's replies above. By moving a window you are stealing the active id away from an input text (or anything else, for that matter), and calling calling `SetKeyboardFocusHere()` (or an hypothetical similar `ActivateItem()`) repeatedly every frame you are preventing any other widgets and actions from happening including moving a window, those two things are stealing a same resource from each other.

TL;DR; Focus != Activation. 
- SetKeyboardFocusHere = focus + activation (misnamed function)
- SetItemDefaultFocus() = focus

Read our messages again, also you haven't answered the important question being "why are you not just using `if (!IsAnyItemActive())` or another trigger" for activating your item.

**Comment by JPGygax68** [comment]
I understand your explanation of what is happening. I am using `if (!IsAnyItemActive())` now. I'll leave the rest to you. Thanks for answering.

**Comment by ocornut** [comment]
Hello @JPGygax68, any update on this?

Because I've been working recently on a semi-large rework toward providing those activation/focus function so hopefully things will become clearer soon.

**Comment by JPGygax68** [comment]
Hi @ocornut, sorry for the late response. No, I haven't worked on that particular issue and your suggestions back then were enough for my needs. I'm looking forward to seeing, and working with your rework of course.
